spec/src/main/asciidoc/configexamples.asciidoc

```diff
@@ -119,3 +119,145 @@ public class InjectedConfigUsageSample {
 }
 ----
 
+=== Aggregate related properties into a single POJO Example
+
+When injecting a number of related configuration properties, it can be tedious to repeat the statement of `ConfigProperty` in scatter places. 
+Since they are related, it makes more sense to aggregate them into a single property class. 
+
+MicroProfile Config provides a way to look up a number of configuration properties starting with the same prefix using the `@ConfigProperties` annotation, e.g. `ConfigProperties(prefix="myPrefix")`.
+When annotating a class with `@ConfigProperties` or `@ConfigProperties(prefix="myPrefix")`, any of its fields, regardless of the visibility, maps to a configuration property via the following mapping rules. 
+
+* If the `prefix` is present, the field `x` maps to the configuration property `<prefix>.x`. 
+* If the `prefix` is absent, the field `x` maps to the property name `x`. 
+
+If the field name `x` needs to be different from the config property name `y`, use `@ConfigProperty(name="y")` to perform the transformation. 
+If the prefix is present, the field `x` maps to the configuration property `<prefix>.y`, otherwise `y`.
+
+Considering the following config sources:
+
+[source, text]
+----
+config_ordinal = 120
+server.host = localhost
+server.port=9080
+server.endpoint=query
+server.old.location=London
+----
+
+[source, text]
+----
+config_ordinal = 150
+client.host = myHost
+client.port=9081
+client.endpoint=shelf
+client.old.location=Dublin
+host = anotherHost
+port=9082
+endpoint=book
+old.location=Berlin
+----
+
+In order to retrieve the above properties in a single property class, you can use the `@ConfigProperties` annotation with a prefix.
+
+[source, java]
+----
+
+@ConfigProperties(prefix="server")
+public class Details {
+    public String host; // the value of the configuration property server.host
+    public int port;   // the value of the configuration property server.port
+    private String endpoint; //the value of the configuration property server.endpoint
+    public @ConfigProperty(name="old.location")
+    String location; //the value of the configuration property server.old.location 
+    public String getEndpoint() {
+        return endpoint;      
+    }     
+}
+----
+The single property class must contain a zero-arg constructor. Otherwise, DeploymentException will be thrown. 
+
+In order to avoid this, you can supply a default value when defining the field. Alternatively, you can use `@ConfigProperty` to provide a default value. 
+You can also use `Optional<T>` or OptionalInt, OptionalDouble, OptionalLong as the type.
+If any of the property cannot be resolved and there is neither default value nor property is not optional, DeploymentException will be thrown. 
```

Comment:
For the programatic API, we cannot throw a `DeploymentException` since it doesn't make sense. Plus, in the way our current API is designed, the class validation can only be performed on the fly, when you call one of the methods to translate into the configuration properties class. This makes me wonder if we should provide a way to provide eager registration and validate.

Reply:
You were right @radcortez. I should place this under the injection lookup instead of here. We should use the exact same handling with `ConfigProperty` and `getValue()`. For getConfigProperties(), it will throw IllegalArgumentException for conversion issue or java.util.NoSuchElementException when the property is missing. For Injection lookup, we can validation the injection point and then throw DeploymentException if there is any problem.

Reply:
@radcortez Based on your feedback, I reworked and hope it makes sense now.